adoptingerlang.org

```diff
@@ -1331,6 +1331,364 @@ Repeat this exercise from time to time, and eventually do some [[https://princip
+
+In general your workflow might look like this:
+
+1. set up the initial project, compile once, and track the lock file in source control
+2. find out you have a top-level dependency you want to change
+3. change the dependency definition in the rebar.config file (or optionally if using a non-versioned hex package, call =rebar3 update=)
+4. call =rebar3 upgrade <app>= to update the application and its transitive dependencies
+
+And that's it, you're done. At least for dependencies you don't maintain; for those, things can get a bit annoying.
+
+*** Checkout Dependencies
+
+If you're just adopting Erlang, whether for fun or at work, chances are you'll have one project to maintain, maybe a few more than that, and small ones at that. Working with a few repositories won't be too hard. But sooner or later, if you're working in a corporate environment with dozens and dozens of repositories, working with rebar3 might become frustrating.
+
+The problem would be that every time you want to try and modify a dependency to do something that its parent application needs, you'll have to commit and publish the change somewhere rebar3 can fetch it. This becomes a huge pain in the butt really fast.
+
+So there's a little trickshot of a feature called _checkout dependencies_. Checkout dependencies work as follows:
+
+- you have your main project somewhere on your file system
+- the dependency is declared in the rebar.config file of the main project
+- you also have that dependency somewhere on your file system, as a standalone project
+- you add a =_checkouts/= directory to the main project
+- you either copy or symlink the dependency's directory in the =_checkouts/= directory
+
+From that point on, every time the main app is built, it will add an =ebin/= directory to the dependency's directory in checkouts, and re-compile it as if it were a top-level application within the main app.
+
+You can then test the changes to your dependency within the main application until they're ready. Once you're done, remove the =ebin/= directory from the dependency, commit and publish your code in the dependency, remove it from the =_checkouts= directory, and =rebar3 upgrade= it.
+
+This lets you do a lot of small iterative changes locally on a dependency, within the context of the main app, without having to make everything public all the time. It massively reduces the overhead of per-app repositories as an overall development strategy.
+
+*** Using Elixir Dependencies
+
+For many years, the Elixir and Erlang road was a one way street. You could call Erlang from Elixir, but the opposite wasn't true. Since then, and thanks to the support of the Erlang Ecosystem Foundation, changes have been made to Rebar3 to give it a brand new compiler management structure. This structure has made it possible to write compiler plugins that can let Erlang users use Elixir code.
```

Comment:
```suggestion
For many years, the Elixir and Erlang road was a one way street. You could include Erlang dependencies in your Elixir project, but the opposite wasn't true. Since then, and thanks to the support of the Erlang Ecosystem Foundation, changes have been made to Rebar3 to give it a brand new compiler management structure. This structure has made it possible to write compiler plugins that can let Erlang users use Elixir code.
```